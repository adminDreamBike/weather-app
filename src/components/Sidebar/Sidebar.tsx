"use client";

import { convertKelvinToCelsius, getDate } from "@/libs/utils";
import { useLocationStore } from "@/store/location";
import { useWeatherStore } from "@/store/weather";
import {
  Box,
  BoxProps,
  Drawer,
  DrawerContent,
  Flex,
  useColorModeValue,
  useDisclosure,
  Text,
  Icon as IconChakra,
  Input,
} from "@chakra-ui/react";
import { IoLocationSharp } from "react-icons/io5";
import { Icon } from "../Icon/Icon";
import { CityContext } from "@/context/city";
import usePersistStore from "@/hooks/usePersistStore";
import { useContext } from "react";

interface SidebarProps extends BoxProps {
  onClose: () => void;
}

const SidebarContent = ({ onClose, ...rest }: SidebarProps) => {
  const { setName } = useContext(CityContext);
  const location =
    usePersistStore(useLocationStore, (state) => state.location) || [];
  const weather = usePersistStore(useWeatherStore, (state) => state.weather);
  const { name } = location[0] || {};
  const { current } = weather || {};
  const { weather: currentWeather = [] } = current || {};
  const { main } = currentWeather[0] || {};

  const handleSearchWeather = (
    event: React.KeyboardEvent<HTMLInputElement>
  ) => {
    if (event.keyCode === 13) {
      setName(event.currentTarget.value);
    }
  };
  return (
    <Box
      bg="#1e213a"
      borderRight="1px"
      borderRightColor="#1e213a"
      w={{ base: "full", md: 60 }}
      pos="fixed"
      h="full"
      {...rest}
    >
      <Flex
        h="20"
        alignItems="center"
        mx="8"
        flexDirection="column"
        my="10"
        height="full"
      >
        <Input
          placeholder="Search for places"
          height="30px"
          color="white"
          onKeyDown={handleSearchWeather}
          spellCheck={true}
        />
        <Flex
          flexDirection="column"
          justifyContent="space-between"
          height="40vh"
          alignItems="center"
          my="8"
        >
          <Icon main={main} size={20} />
          <Text color="#9ea6bf" fontSize="xxx-large" fontWeight="bold">
            {convertKelvinToCelsius(current?.temp)}
            <Text as="sup" fontWeight="normal">
              c
            </Text>
          </Text>
          <Text color="#9ea6bf" fontSize="x-large">
            {main}
          </Text>
          <Flex flexDirection="row" gap="5">
            <Text color="#9ea6bf">Today </Text>
            <Text color="#9ea6bf">{getDate()}</Text>
          </Flex>
          <Flex flexDirection="row" gap="2">
            <IconChakra as={IoLocationSharp} w={5} h={5} color="#a8a7b8" />
            <Text color="#88879a">{name}</Text>
          </Flex>
        </Flex>
      </Flex>
    </Box>
  );
};
export const Sidebar = () => {
  const { isOpen, onClose } = useDisclosure();

  return (
    <Box bg={useColorModeValue("gray.100", "gray.900")}>
      <SidebarContent
        onClose={() => onClose}
        display={{ base: "none", md: "block" }}
      />
      <Drawer
        isOpen={isOpen}
        placement="left"
        onClose={onClose}
        returnFocusOnClose={false}
        onOverlayClick={onClose}
        size="full"
      >
        <DrawerContent>
          <SidebarContent onClose={onClose} />
        </DrawerContent>
      </Drawer>
    </Box>
  );
};
