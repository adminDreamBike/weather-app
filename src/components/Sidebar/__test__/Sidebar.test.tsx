import { render } from "@testing-library/react";
import { Sidebar } from "../Sidebar";
import { ChakraUIProvider } from "@/providers/ChakraUIProvider";

describe("Sidebar component", () => {
  test("", () => {
    const { asFragment } = render(
      <ChakraUIProvider>
        <Sidebar />
      </ChakraUIProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
