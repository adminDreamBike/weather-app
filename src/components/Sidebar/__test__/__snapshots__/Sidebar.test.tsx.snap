// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar component >  1`] = `
<DocumentFragment>
  <div
    class="css-7luvuw"
  >
    <div
      class="css-16mn4m5"
    >
      <div
        class="css-r2hi9e"
      >
        <input
          class="chakra-input css-vn5z69"
          placeholder="Search for places"
          spellcheck="true"
        />
        <div
          class="css-4hu1sv"
        >
          <div
            class="css-lgj0h8"
          >
            <svg
              class="chakra-icon css-16c8gqk"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.473 9a4.5 4.5 0 0 0-8.72-.99A3 3 0 0 0 3 14h8.5a2.5 2.5 0 1 0-.027-5"
              />
              <path
                d="M14.544 9.772a3.5 3.5 0 0 0-2.225-1.676 5.5 5.5 0 0 0-6.337-4.002 4.002 4.002 0 0 1 7.392.91 2.5 2.5 0 0 1 1.17 4.769z"
              />
            </svg>
          </div>
          <p
            class="chakra-text css-1l97044"
          >
            25.1
            <sup
              class="chakra-text css-11cuipc"
            >
              c
            </sup>
          </p>
          <p
            class="chakra-text css-aqyol4"
          >
            Clouds
          </p>
          <div
            class="css-e8tbaw"
          >
            <p
              class="chakra-text css-mp8c6w"
            >
              Today 
            </p>
            <p
              class="chakra-text css-mp8c6w"
            >
              Tue, 22 Oct
            </p>
          </div>
          <div
            class="css-1fb3o3p"
          >
            <svg
              class="chakra-icon css-bv7m74"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 32C167.67 32 96 96.51 96 176c0 128 160 304 160 304s160-176 160-304c0-79.49-71.67-144-160-144zm0 224a64 64 0 1 1 64-64 64.07 64.07 0 0 1-64 64z"
              />
            </svg>
            <p
              class="chakra-text css-1mlu7ca"
            >
              Medellin
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
